### PR TITLE
Migrate `CalculateBallisticAcceleration.lua` to `defaultweapons.lua`

### PR DESCRIPTION
**Pull request by Hdt80bro**
Addresses [3739](https://github.com/FAForever/fa/issues/3739) (yes I tested this--the results are the same)
As requested, the byte code. It's longer because the `MuzzleSalvoSize==1` case has been optimized to not cache or include salvo logic.
```
(  42)    0 - GETUPVAL       2    0    0
(  42)    1 - MOVE           3    1    0
(  42)    2 - CALL           2    2    2
(  43)    3 - TEST           2    2    1
(  43)    4 - JMP            0    2
(  44)    5 - LOADK          3    0
(  44)    6 - RETURN         3    2    0
(  46)    7 - GETUPVAL       3    1    0
(  47)    8 - GETTABLE       4    2  251
(  51)    9 - GETUPVAL       5    2    0
(  51)   10 - MOVE           6    1    0
(  51)   11 - CALL           5    2    4
(  52)   12 - GETUPVAL       8    3    0
(  52)   13 - MOVE           9    2    0
(  52)   14 - CALL           8    2    4
(  54)   15 - GETUPVAL      11    4    0
(  54)   16 - MOVE          12    2    0
(  54)   17 - CALL          11    2    2
(  56)   18 - LOADNIL       12   16    0
(  59)   19 - TEST          11   11    0
(  59)   20 - JMP            0   18
(  59)   21 - GETUPVAL      17    5    0
(  59)   22 - MOVE          18   11    0
(  59)   23 - CALL          17    2    2
(  59)   24 - TEST          17   17    0
(  59)   25 - JMP            0   13
(  61)   26 - GETUPVAL      17    2    0
(  61)   27 - MOVE          18   11    0
(  61)   28 - CALL          17    2    4
(  61)   29 - MOVE          14   19    0
(  61)   30 - MOVE           9   18    0
(  61)   31 - MOVE          13   17    0
(  62)   32 - GETUPVAL      17    3    0
(  62)   33 - MOVE          18   11    0
(  62)   34 - CALL          17    2    4
(  62)   35 - MOVE          16   19    0
(  62)   36 - MOVE           9   18    0
(  62)   37 - MOVE          15   17    0
(  62)   38 - JMP            0   10
(  65)   39 - GETUPVAL      17    6    0
(  65)   40 - MOVE          18    0    0
(  65)   41 - CALL          17    2    2
(  65)   42 - MOVE          12   17    0
(  66)   43 - GETTABLE      17   12  252
(  66)   44 - GETTABLE      14   12  253
(  66)   45 - MOVE          13   17    0
(  67)   46 - LOADK         17    4
(  67)   47 - LOADK         16    4
(  67)   48 - MOVE          15   17    0
(  70)   49 - GETTABLE      17    3    4
(  73)   50 - TEST          17   17    1
(  73)   51 - JMP            0   78
(  74)   52 - GETTABLE      18    0  255
(  75)   53 - GETTABLE      19   18  256
(  77)   54 - LT             0  252   19
(  77)   55 - JMP            0   24
(  79)   56 - NEWTABLE      20    0    3
(  80)   57 - SETTABLE      20  257   19
(  81)   58 - SETTABLE      20  258  250
(  84)   59 - MUL           21  260   19
(  84)   60 - SUB           21   21  260
(  84)   61 - SETTABLE      20  259   21
(  87)   62 - GETTABLE      21   18  263
(  87)   63 - MUL           21  262   21
(  87)   64 - SETTABLE      20  261   21
(  89)   65 - SETTABLE      20  264   12
(  90)   66 - MOVE          17   20    0
(  91)   67 - GETTABLE      20   18  265
(  92)   68 - TEST          20   20    0
(  92)   69 - JMP            0    8
(  94)   70 - GETUPVAL      21    7    0
(  94)   71 - LOADK         22    4
(  94)   72 - GETUPVAL      23    8    0
(  94)   73 - SUB           24  252   20
(  94)   74 - LOADK         25    2
(  94)   75 - CALL          23    3    0
(  94)   76 - CALL          21    0    2
(  94)   77 - SETTABLE      17  266   21
(  96)   78 - SETTABLE       3    4   17
(  96)   79 - JMP            0   50
(  99)   80 - GETTABLE      20   11  267
(  99)   81 - TEST          20   20    0
(  99)   82 - JMP            0    2
( 100)   83 - LOADK         20    0
( 100)   84 - RETURN        20    2    0
( 102)   85 - GETUPVAL      20    9    0
( 102)   86 - MOVE          21    8    0
( 102)   87 - MOVE          22   10    0
( 102)   88 - MOVE          23   15    0
( 102)   89 - MOVE          24   16    0
( 102)   90 - CALL          20    5    2
( 103)   91 - EQ             0   20  254
( 103)   92 - JMP            0    2
( 104)   93 - LOADK         21    0
( 104)   94 - RETURN        21    2    0
( 106)   95 - GETUPVAL      21    9    0
( 106)   96 - MOVE          22    5    0
( 106)   97 - MOVE          23    7    0
( 106)   98 - MOVE          24   13    0
( 106)   99 - MOVE          25   14    0
( 106)  100 - CALL          21    5    2
( 107)  101 - GETTABLE      22   18  265
( 108)  102 - TEST          22   22    0
( 108)  103 - JMP            0    8
( 109)  104 - GETUPVAL      23    7    0
( 109)  105 - LOADK         24    4
( 109)  106 - GETUPVAL      25    8    0
( 109)  107 - SUB           26  252   22
( 109)  108 - LOADK         27    2
( 109)  109 - CALL          25    3    0
( 109)  110 - CALL          23    0    2
( 109)  111 - MUL           21   21   23
( 111)  112 - EQ             0   21  254
( 111)  113 - JMP            0    2
( 112)  114 - LOADK         23    0
( 112)  115 - RETURN        23    2    0
( 114)  116 - DIV           23   21   20
( 115)  117 - MUL           24   23   15
( 115)  118 - ADD           24   13   24
( 116)  119 - MUL           25   23   16
( 116)  120 - ADD           25   14   25
( 117)  121 - GETUPVAL      26   10    0
( 117)  122 - MOVE          27   24    0
( 117)  123 - MOVE          28   25    0
( 117)  124 - CALL          26    3    2
( 118)  125 - SUB           27    6   26
( 118)  126 - MUL           27  268   27
( 118)  127 - MUL           28   23   23
( 118)  128 - DIV           27   27   28
( 118)  129 - RETURN        27    2    0
( 124)  130 - GETTABLE      18   11  267
( 124)  131 - TEST          18   18    1
( 124)  132 - JMP            0    3
( 124)  133 - GETTABLE      18   17  269
( 124)  134 - TEST          18   18    0
( 124)  135 - JMP            0   10
( 126)  136 - GETTABLE      18   17  257
( 126)  137 - SUB           18   18  252
( 127)  138 - SETTABLE      17  257   18
( 128)  139 - LOADBOOL      19    1    0
( 128)  140 - SETTABLE      17  269   19
( 129)  141 - EQ             0   18  254
( 129)  142 - JMP            0    1
( 130)  143 - SETTABLE       3    4  270
( 132)  144 - GETTABLE      19   17  258
( 132)  145 - RETURN        19    2    0
( 137)  146 - GETUPVAL      18    9    0
( 137)  147 - MOVE          19    8    0
( 137)  148 - MOVE          20   10    0
( 137)  149 - MOVE          21   15    0
( 137)  150 - MOVE          22   16    0
( 137)  151 - CALL          18    5    2
( 138)  152 - EQ             0   18  254
( 138)  153 - JMP            0    2
( 139)  154 - LOADK         19    0
( 139)  155 - RETURN        19    2    0
( 141)  156 - GETUPVAL      19    9    0
( 141)  157 - MOVE          20    5    0
( 141)  158 - MOVE          21    7    0
( 141)  159 - MOVE          22   13    0
( 141)  160 - MOVE          23   14    0
( 141)  161 - CALL          19    5    2
( 143)  162 - GETTABLE      20   17  266
( 144)  163 - TEST          20   20    0
( 144)  164 - JMP            0    1
( 145)  165 - MUL           19   19   20
( 149)  166 - GETTABLE      21   17  257
( 149)  167 - SUB           21   21  252
( 153)  168 - GETTABLE      22   17  261
( 153)  169 - MUL           22   22   18
( 153)  170 - GETTABLE      23   17  259
( 153)  171 - SUB           23   23   21
( 153)  172 - MUL           22   22   23
( 153)  173 - ADD           19   19   22
( 154)  174 - SETTABLE      17  257   21
( 155)  175 - EQ             0   21  254
( 155)  176 - JMP            0    1
( 156)  177 - SETTABLE       3    4  270
( 158)  178 - EQ             0   19  254
( 158)  179 - JMP            0    2
( 159)  180 - LOADK         22    0
( 159)  181 - RETURN        22    2    0
( 163)  182 - DIV           22   19   18
( 166)  183 - MUL           23   22   15
( 166)  184 - ADD           23   13   23
( 167)  185 - MUL           24   22   16
( 167)  186 - ADD           24   14   24
( 169)  187 - GETUPVAL      25   10    0
( 169)  188 - MOVE          26   23    0
( 169)  189 - MOVE          27   24    0
( 169)  190 - CALL          25    3    2
( 178)  191 - SUB           26    6   25
( 178)  192 - MUL           26  268   26
( 178)  193 - MUL           27   22   22
( 178)  194 - DIV           26   26   27
( 181)  195 - SETTABLE      17  258   26
( 182)  196 - RETURN        26    2    0
( 183)  197 - RETURN         0    1    0
maxstack = 29
```

### CODE REVIEW

**Comment by Garanas** [comment]
@MrRowey  / @Spikey84 could either of you test whether bombers still work as they used to do with these changes? The idea is that the projectiles they generate act the same as before, within reason of course. From a code perspective (no more unnecessary table allocations, better readable code, etc) it looks good!

**Comment by MrRowey** [comment]
Will test when I'm home from work 🤠

**Comment by Spikey84** [comment]
Would this only effect projectile or also bomber movement? If only projectiles it seems perfectly fine. But i noticed some changes when trying to hover bomb, this could totally just be a fluke and meaningless but i figured i should bring it up.

**Comment by Hdt80bro** [comment]
It *should* only affect the projectiles. But if you're seeing a difference with the bomber movement, we should definitely look into that!

**Comment by Spikey84** [comment]
Chances are it is just a fluke, but maybe someone else can try hover bombing and make sure it works as normal.

**Comment by Garanas** [comment]
The results are not identical.

```
Output of the Janus, the method introduced by this pull request:

INFO: New result: 6
INFO: New result: 6
INFO: New result: 6
INFO: New result: 6
INFO: New result: 6
INFO: New result: 6
INFO: New result: 7
INFO: New result: 7
INFO: New result: 7
INFO: New result: 7
INFO: New result: 6507
INFO: New result: 7149
INFO: New result: 6322
INFO: New result: 6240
INFO: New result: 8860
INFO: New result: 892
INFO: New result: 144
INFO: New result: 50
INFO: New result: 22
INFO: New result: 12
```

```
Output of the Janus, previously
INFO: Old result: 7
INFO: Old result: 8
INFO: Old result: 9
INFO: Old result: 10
INFO: Old result: 12
INFO: Old result: 14
INFO: Old result: 16
INFO: Old result: 18
INFO: Old result: 21
INFO: Old result: 25
INFO: Old result: 29
INFO: Old result: 36
INFO: Old result: 44
INFO: Old result: 55
INFO: Old result: 74
INFO: Old result: 73
INFO: Old result: 48
INFO: Old result: 29
INFO: Old result: 17
INFO: Old result: 12
```

It appears that bombers with just one bomb drop are identical, but bombers with multiple are not. With that said, I think the current solution (using bomb data) is bad, as it introduces quite a bit of global scope while we can also store that in the weapon instance, keeping the function pure to its parameters.

What are your thoughts on it?

**Comment by Garanas** [comment]
Note that to make the comparison for a bomber with multiple bombers you need to comment out one of the LOGs, as they influence one another.

**Comment by Hdt80bro** [comment]
Interesting, I'll take a look

**Comment by Garanas** [comment]
And this does **not** affect bomber movement - this only influences projectiles that are already being dropped.

**Comment by Hdt80bro** [comment]
Moving the data to the weapon sounds a lot better, though it might mean a few more checks to make sure the weapon is valid. Shouldn't be too bad, and it'd also give us more control on the weapon side of things.

**Comment by Garanas** [comment]
I agree, why isn't this function part of the weapon anyway?

**Comment by Garanas** [comment]
As in, part of the weapon class?

**Comment by Hdt80bro** [comment]
I... couldn't tell you. I'm suddenly having lots of refactoring ideas for this function

**Comment by Garanas** [comment]
Go at it. The numbers are slightly different, but it isn't notable while playing.

**Comment by Garanas** [comment]
The function is used by some mods however, see also:

image

What is the best approach here? These are two active mods. All the other large mod packs do not reference this function, including BlackOps, Marlo's Mod pack and TotalMayhem.

**Comment by Garanas** [comment]
I think we should tell Maudlin / Relent0r to refactor their approach, hook the weapon class with the new function.

**Comment by Hdt80bro** [comment]
Ohhhh, that's right--mod compatiblity. Yeah, we can leave the function in for now while we move it to the weapon class, and ask the mod devs to refactor.

**Comment by Hdt80bro** [comment]
I'll setup the framework first though so we know what we're doing

**Comment by Hdt80bro** [comment]
Something interesting to note is that some code has begun to rely on the global table, such as [this](https://github.com/FAForever/fa/blob/deploy/fafdevelop/lua/terranweapons.lua#L244) (Again, something easily solved by moving the data to the weapon). I'm worried that some mods have begun to do the same.

**Comment by Hdt80bro** [comment]
Hmm, that might have been a little aggressive of an optimization for this branch.
Well, the important bits are that:
 * [`Weapon.CurrentSalvoNumber` gets set in between shots does](https://github.com/Hdt80bro/fa/blob/optimize-calculate-trajectories/lua/sim/defaultweapons.lua#L970), like `Weapon.CurrentRackSalvoNumber` does
 * The calculation has been [moved to `defaultweapon.lua`](https://github.com/Hdt80bro/fa/blob/optimize-calculate-trajectories/lua/sim/defaultweapons.lua#L164-L176), and the one [side effect](https://github.com/Hdt80bro/fa/blob/optimize-calculate-trajectories/lua/terranweapons.lua#L242-L253) taken care of
 * [`CalculateBallisticAcceleration.lua`](https://github.com/Hdt80bro/fa/blob/optimize-calculate-trajectories/lua/sim/CalcBallisticAcceleration.lua) has been deprecated

File: units/DEA0202/DEA0202_unit.bp [hunk 1]
```diff
@@ -304,7 +304,7 @@ UnitBlueprint {
             MaxRadius = 30,
             MuzzleSalvoDelay = 0,
             MuzzleSalvoSize = 1,
-            MuzzleVelocity = 120,
```
Comment (Garanas) [inline units/DEA0202/DEA0202_unit.bp:307]: Was this on purpose?

Comment (Hdt80bro) [reply units/DEA0202/DEA0202_unit.bp:307]: Uh, that was you

File: lua/sim/weapon.lua [hunk 2]
```diff
@@ -104,264 +129,286 @@ Weapon = Class(moho.weapon_methods) {
...
                     end
                 end
             end
         else
             error('*ERROR: Trying to setup a turreted weapon but there are yaw bones, pitch bones or muzzle bones missing from the blueprint.', 2)
         end
+        self.AimControl = aimControl
 
-        local numbersexist = true
+        local numbersExist = true
```
Comment (Garanas) [inline lua/sim/weapon.lua:203]: I always read this as one word 😄 , instead of two

**Comment by Garanas** [comment]
@maudlin27 / @relent0r your AI mods depend on the function `CalculateBallisticAcceleration`. In this PR we migrate that function into the `DefaultProjectileWeapon` class, part of `defaultweapons.lua`. You can hook the function there and add your logic accordingly.

The release is on the 20th - can you both make that in terms of having a release yourself right after?

**Comment by Garanas** [comment]
@Hdt80bro once they've both responded you can merge this accordingly.

**Comment by relent0r** [comment]
Yea all good. When you said the 20th did you mean Aug or Sept?

File: lua/sim/weapon.lua (hunk 2 above)
Comment (Hdt80bro) [reply lua/sim/weapon.lua:203]: Yeah, that had to change :)

**Comment by maudlin27** [comment]
Am I right in thinking that if we wanted to get the location that a bomb will land at, under the new approach we'd hook CreateProjectileAtMuzzle and get the position per data.targetpos? if so should be fine to introduce for 20th.

**Comment by Garanas** [comment]
I'd highly recommend you to use that method - yes! By hooking the weapon you know exactly what the weapon is trying to do, you can even get the actual target.

File: lua/sim/CalcBallisticAcceleration.lua [hunk 3]
```diff
@@ -1,98 +1,4 @@
-local XZDist = import('/lua/utilities.lua').XZDistanceTwoVectors
-
--- This table stores last acceleration and numbers of bombs left in a cluster bomb run, as well as the original target
--- format : bomb_data[entityId] = {n_left=<n_left>, acc=<last_acc>, targetpos=<original_targetpos>}
-bomb_data = {}
-
+---@deprecated
 CalculateBallisticAcceleration = function(weapon, projectile)
```
Comment (Garanas) [inline lua/sim/CalcBallisticAcceleration.lua:2]: Is this still being called?

File: units/DEA0202/DEA0202_unit.bp [hunk 4]
```diff
@@ -453,10 +453,11 @@ UnitBlueprint {
             FiringRandomness = 0.1,
             FiringTolerance = 6,
             FixBombTrajectory = true,
+            DropBombShort = -0.1,
             InitialDamage = 20, -- Total impact damage = 400. Total overall damage = 1600
             Label = 'Bomb',
             MaxRadius = 60,
-            MuzzleSalvoDelay = 0.2,
+            MuzzleSalvoDelay = 0.1,
```
Comment (Garanas) [inline units/DEA0202/DEA0202_unit.bp:460]: These changes are intended, prevents the Janus from firing backwards

File: lua/sim/CalcBallisticAcceleration.lua (hunk 3 above)
Comment (Hdt80bro) [reply lua/sim/CalcBallisticAcceleration.lua:2]: Not in our code. I don't know how many other mods there out there that do though

**Comment by Hdt80bro** [comment]
> Am I right in thinking that if we wanted to get the location that a bomb will land at, under the new approach we'd hook CreateProjectileAtMuzzle and get the position per data.targetpos? if so should be fine to introduce for 20th.

Actually... not currently. It only populates that field for groundfires at the moment, but it'd be easy enough to change that to support modding.

**Comment by Hdt80bro** [comment]
Alright, I've changed it to always update `targetpos`

File: units/DEA0202/DEA0202_unit.bp (hunk 4 above)
Comment (Hdt80bro) [reply units/DEA0202/DEA0202_unit.bp:460]: How does that work?

**Comment by Hdt80bro** [comment]
Is there any other information we'd like in `CurrentSalvoData`?

File: lua/sim/CalcBallisticAcceleration.lua (hunk 3 above)
Comment (Garanas) [reply lua/sim/CalcBallisticAcceleration.lua:2]: My mod folder is quite large, none of them appear to do so

File: units/DEA0202/DEA0202_unit.bp (hunk 4 above)
Comment (Garanas) [reply units/DEA0202/DEA0202_unit.bp:460]: By releasing the projectiles quicker. Previously some projectiles would always be released after the Janus already passed its target. Now they all release before it does so

File: lua/sim/CalcBallisticAcceleration.lua (hunk 3 above)
Comment (Hdt80bro) [reply lua/sim/CalcBallisticAcceleration.lua:2]: Alright, well, it should be currently marked for removal--if you think we can skip that and just delete it, go for it.

**Comment by Garanas** [comment]
What type of information would fit in there? A lot of the information can already be found in the weapon itself, for example

File: units/DEA0202/DEA0202_unit.bp (hunk 4 above)
Comment (Hdt80bro) [reply units/DEA0202/DEA0202_unit.bp:460]: Gotcha

**Comment by Hdt80bro** [comment]
Stuff like projected land location (which might be what `targetpos` should already be--see newTargetPos), bombs remaining (which can be pulled out of the new `CurrentSalvoNumber` field), estimated drop time...

**Comment by Hdt80bro** [comment]
It might be worth pointing out the new fields that get set when `FixBombTrajectory` is set
If `DropBombShort` is set:
 * `DropBombShortRatio` is the fraction of the horizontal distance between the projectile and target position used in the calculation

if `MuzzleSalvoSize` is greater than one, then:
 * `SalvoSpreadStart` is the number that centers the bomb clusters around the optimal release position. If you're curious, the formula is `SalvoSpreadStart = -(1 + MuzzleSalvoSize) / 2` and the relevant calculation is `distance = distance + AdjustedSalvoDelay * velocityDifference * (SalvoSpreadStart + CurrentSalvoNumber)`
 * `AdjustedSalvoDelay` is how many ticks between bombs, "multiplied by 0.5 to get the bombs overlapping a bit"

File: units/DEA0202/DEA0202_unit.bp [hunk 5]
```diff
@@ -453,10 +453,11 @@ UnitBlueprint {
             FiringRandomness = 0.1,
             FiringTolerance = 6,
             FixBombTrajectory = true,
+            DropBombShort = -0.1,
```
Comment (Hdt80bro) [inline units/DEA0202/DEA0202_unit.bp:456]: By the way, this gets clamped, so going below 0 doesn't make it release beyond its optimal location--at that point it's better to not include the line.

Comment (Garanas) [reply units/DEA0202/DEA0202_unit.bp:456]: I recall it did change things, allow me to check again.

Comment (Garanas) [reply units/DEA0202/DEA0202_unit.bp]: You were correct 👍

**Comment by Garanas** [comment]
Merging this in, we can add more information in the future if required. But I'll be honest - I see no immediate need for this

**Comment by maudlin27** [comment]
It's quite possible this is due to an error in my AI rather than the newly added code, but sometimes I get the following error - I've included reprs of the variables relating to this, with the projectile being nil:

info: CalculateBallisticAcceleration: reprs of projectile=
info: Printing information of projectile of type: /projectiles/cifneutronclusterbomb01/cifneutronclusterbomb01_proj.bp, and with entity id: nil
info:  - Army: 2
info:  - Blueprint: table: 1AE260C8 (skipped) 
info:  - DamageData: table: 272FCE38 (skipped) 
info:  - Impacted: false
info:  - Launcher: table: 28A6D2D0 (Unit of type: ura0103, with entity id: 1048595) 
info:  - OriginalTarget: table: 1956C488 (Unit of type: uel0201, with entity id: 2097187) 
info:  - Trash: table: 272FC758 (Trashbag) 
info:  - _c_object: userdata: CScriptObject* at 56372B8C = [CScriptObject at 0x5773EA80]
info: CalculateBallisticAcceleration: reprs of self=
info: Printing information of state with identifier: 25
info:  - AdjustedSalvoDelay: 1
info:  - Army: 2
info:  - Blueprint: table: 1A8FA8E8 (skipped) 
info:  - Brain: table: 1E87EBE0 (Brain of army: 2) 
info:  - CollideFriendly: false
info:  - CurrentRackSalvoNumber: 1
info:  - CurrentSalvoData: table: 2698DAF0
info:  -    lastAccel: 5.6753835678101
info:  -    targetpos: table: 199319B0 (Vector { 484.18280029297, 22.0080909729, 224.96257019043}) 
info:  - CurrentSalvoNumber: 5
info:  - DisabledBuffs: table: 23E83988
info:  - FirstShot: false
info:  - HaltFireOrdered: false
info:  - Label: Bomb
info:  - NumMuzzles: 6
info:  - NumRackBones: 1
info:  - SalvoSpreadStart: -3.5
info:  - Trash: table: 23E83DE8 (Trashbag) 
info:  - TrashManipulators: table: 28A6DD48 (Trashbag) 
info:  - TrashProjectiles: table: 271A0D48 (Trashbag) 
info:  - WeaponAimIsEnabled: true
info:  - WeaponCanFire: true
info:  - WeaponIsEnabled: true
info:  - __mainthread: thread: 1EE4AC80
info:  - _c_object: userdata: CScriptObject* at 1F2B3CCC = [CScriptObject at 0x56D13C40]
info:  - damageTableCache: table: 268493C0
info:  -    Buffs: table: 26849910
info:  -    CollideFriendly: false
info:  -    DamageAmount: 50
info:  -    DamageFriendly: true
info:  -    DamageRadius: 3
info:  -    DamageType: Normal
info:  -    InitialDamageAmount: 0
info:  - unit: table: 28A6D2D0 (Unit of type: ura0103, with entity id: 1048595) 
info: CalculateBallisticAcceleration: reprs of projectile=
info: Printing information of projectile of type: /projectiles/cifneutronclusterbomb01/cifneutronclusterbomb01_proj.bp, and with entity id: nil
info:  - Army: 2
info:  - Blueprint: table: 1AE260C8 (skipped) 
info:  - DamageData: table: 1F545938 (skipped) 
info:  - Impacted: false
info:  - Launcher: table: 28A6D2D0 (Unit of type: ura0103, with entity id: 1048595) 
info:  - Trash: table: 1F545960 (Trashbag) 
info:  - _c_object: userdata: CScriptObject* at 5663A498 = [CScriptObject at 0x55F59C00]
warning: Error running lua script: attempt to use as number a nil value
         stack traceback:
         	[C]: ?
         	...aforever\gamedata\lua.nx5\lua\sim\defaultweapons.lua(248): in function <...aforever\gamedata\lua.nx5\lua\sim\defaultweapons.lua:170>
         	(tail call): ?
         	...aforever\gamedata\lua.nx5\lua\sim\defaultweapons.lua(167): in function `CheckBallisticAcceleration'
         	...aforever\gamedata\lua.nx5\lua\sim\defaultweapons.lua(158): in function `CreateProjectileAtMuzzle'
         	...aforever\gamedata\lua.nx5\lua\sim\defaultweapons.lua(879): in function <...aforever\gamedata\lua.nx5\lua\sim\defaultweapons.lua:793>

**Comment by maudlin27** [comment]
My hook of it (that resulted in the below log) is:
https://github.com/maudlin27/M27AI/blob/main/hook/lua/sim/defaultweapons.lua

**Comment by Garanas** [comment]
@Hdt80bro do you have the time to investigate?

**Comment by maudlin27** [comment]
[2022-08-15 M27AI-48-main wip.zip](https://github.com/FAForever/fa/files/9336144/2022-08-15.M27AI-48-main.wip.zip)
[BallisticBugReplay.zip](https://github.com/FAForever/fa/files/9336148/BallisticBugReplay.zip)

I've attached the replay and version of M27AI used in the game where it arose, I think its caused by one of RNG's bombers between gametime 407-417.  Vault versions of RNG and Uveso are required as well, and I was using FAF develop that I downloaded about 13 hours ago.

**Comment by maudlin27** [comment]
Note I dont normally get this error - e.g. I tried to replicate by just spawning a bunch of strikers for M27 and having Zeus bombers attack them but didnt get the error

**Comment by maudlin27** [comment]
Was hilarious to see but not something desired - my AI built a scathis and every shot landed precisely at the target
image

**Comment by Garanas** [comment]
Hot damn, how about the other artillery 😆

**Comment by Hdt80bro** [comment]
Wait--artillery has its ballistic acceleration fixed as well?

**Comment by Hdt80bro** [comment]
> @Hdt80bro do you have the time to investigate?

I'm looking right now

**Comment by Hdt80bro** [comment]
I'm also finding that the reload bar for energy-consuming units (i.e. artillery) is not synced to how much energy it has consumed.

**Comment by Hdt80bro** [comment]
> warning: Error running lua script: attempt to use as number a nil value

@maudlin27 What's happening is that the acceleration calculation is being called with neither a target unit nor a target location. So I'm not sure how the attack order even went through.

It actually brings to my attention a similar possible problem that could happen when targeting a non-unit entity, because line 185 `if target and IsUnit(target) then` checks only for units, rather than letting in any old entity (like it merely needs). I'm not sure how a non-unit entity would get targeted though. 

Anyway, in this case the target is also `nil`, so it's not that that's happening.

**Comment by Hdt80bro** [comment]
Ohhh, I see, it was updating the `targetpos` even once it had retargeted, making it `nil`. It should be fixed now.

**Comment by Hdt80bro** [comment]
@Garanas only the Scathis had this problem...

You want to guess the issue?

It's truly, truly awful.

Examine this line
```lua
local randomness = bp.FixedSpreadRadius / (distance^2 / 12)
```
What do we think it simplifies to? If you guessed
```lua
local randomness = 12 * bp.FixedSpreadRadius / (distance * distance)
```
like I did, then you're dead wrong

No, the correct answer is
```lua
local randomness = 12 * bp.FixedSpreadRadius 
distance = math.floor(distance)
if distance < 0 then
    randomness = randomness / (4294967296)
elseif math.mod(distance, 4) < 2 then
    randomness = randomness / (distance + 2)
else
    randomness = randomness / (distance - 2)
end
```
Since the overridden `__pow` function gets `^` behavior like bitwise XOR

**Comment by Hdt80bro** [comment]
It also appears that `n ^ 0` is faster than even a localized `math.floor(n)` function call by several orders of magnitude, so at least we found that out.

**Pull request merged**